design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,965 @@
+specifying its `writeConnectionSecretToRef` field.
+
+This document proposes that `AggregateResource` connection secrets contain the
+intersection of the connection secrets of the managed resources they aggregate. 
+For example if an `AggregateResource` matched two managed resources with the
+following secrets:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: secret-a
+stringData:
+  username: cooluser
+  password: verysecure
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: secret-b
+stringData:
+  url: https://example.org
+```
+
+The `AggregateResource` connection secret would be:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: aggregate-secret
+stringData:
+  username: cooluser
+  password: verysecure
+  url: https://example.org
+```
+
+An undefined connection secret key will win in the case of a conflict. Conflicts
+are expected to be rare in practice (most "supporting" managed resources do not
+create a connection secret), but may be avoided by respecting a new
+`aggregation.crossplane.io/connection-secret-prefix` annotation. This annotation
+would be propagated from resource classes to the managed resources they
+dynamically provision, specifying a string to be prepended to their connection
+secret keys during aggregation.
+
+## User Experience
+
+Consider an infrastructure operator who wishes to enable their application
+operator peers to submit `MySQLInstance` resource claims and have them be
+satisfied by dynamically provisioned Azure `MySQLServer` managed resources. Each
+`MySQLServer` should allow traffic from a statically provisioned subnetwork
+using a `MySQLServerVirtualNetworkRule`. Each dynamically provisioned
+`MySQLServer` and its associated virtual network rule should be created in its
+own resource group.
+
+First the infrastructure operator statically provisions the `Subnet` from which
+all dynamically provisioned `MySQLServer` managed resources should allow
+traffic:
+
+```yaml
+apiVersion: network.azure.crossplane.io/v1alpha3
+kind: Subnet
+metadata:
+  name: private
+spec:
+  resourceGroupName:
+    name: some-existing-group
+  virtualNetworkName:
+    name: some-existing-network
+  properties:
+    addressPrefix: 10.2.0.0/24
+    serviceEndpoints:
+      - service: Microsoft.Sql
+  providerRef:
+    name: example
+```
+
+The infrastructure operator then authors resource classes for each of the
+managed resources they wish to allow application operators to dynamically
+provision, and an aggregate resource class to aggregate them:
+
+```yaml
+apiVersion: aggregation.crossplane.io/v1alpha1
+kind: AggregateResourceClass
+metadata:
+  name: private-mysql-server
+  annotations:
+    # This will be the default resource class for any resource claim capable of
+    # binding directly to a resource of one of the classes it aggregates - in
```

Comment:
>  I am not sure that I feel great about being able to use a MySQLInstance or a KubernetesCluster to provision a sql-backed-gke

This would not actually be possible - I've added a paragraph that clarifies this. In practice the set of classes deemed to match the `AggregateResourceClass` would depend on which classes were compatible with the underlying claim.

I do like the idea of better documenting which claims may be satisfied by which resource classes, though with Crossplane's current architecture in which those relationships are defined at the controller level this would by necessity document rather than dictate.